### PR TITLE
fix(refit): release trainer memory before collective export

### PR DESCRIPTION
**Pull request by seonjinn**
# What does this PR do?

Adds opt-in trainer-memory release before non-colocated vLLM refit using either the default NCCL collective transport or `nccl_reshard`.

This prevents large MXFP8 exports, including Qwen3.5 397B, from competing with completed gradient buffers, optimizer state, and Transformer Engine workspaces for trainer GPU memory.

The implementation reuses `policy.offload_before_refit()`, supports GRPO and distillation, preserves the existing behavior by default, and rejects unsupported backend, topology, and transport combinations.

# Issues

N/A

# Usage

```yaml
policy:
  release_grads_before_refit: true
  offload_optimizer_for_refit: true
  megatron_cfg:
    fp8_cfg:
      enabled: true
      force_clear_fp8_caches: true
  generation:
    backend: vllm
    colocated:
      enabled: false
    refit_transport: nccl_reshard  # null also supports this option
```

# Before your PR is "Ready for review"

**Pre checks**:

- [x] Read and followed the contributor guidelines.
- [x] Added focused unit coverage for validation, factory propagation, and refit ordering.
- [ ] Run the full unit and functional suites locally. The lockfile supports Linux only; targeted lifecycle harness, Ruff, formatting, compile, and diff checks passed on macOS.
- [x] Updated the refit guide and exemplar/reference configurations.

# Additional Information

The option is disabled by default and requires Megatron policy training with non-colocated vLLM generation. Both the default NCCL collective transport and `nccl_reshard` are supported.

### CODE REVIEW

File: nemo_rl/models/policy/__init__.py [hunk 1]
```diff
@@ -600,6 +600,13 @@ class PolicyConfig(TypedDict):
     # This sets the clipping norm for the DTensorPolicyWorkers (Megatron's is called clip_grad)
     max_grad_norm: NotRequired[float | int | None]
     refit_buffer_size_gb: NotRequired[float | int]
+    # Release completed gradient buffers before non-colocated collective refit.
+    # Existing optimizer and FP8 cache knobs control the additional memory released
+    # by the same offload_before_refit lifecycle.
+    release_grads_before_refit: NotRequired[bool]
```
Comment (yuki-97) [inline nemo_rl/models/policy/__init__.py]: `release_grads_before_refit` is declared on the shared `PolicyConfig`, but it is only honored where a `CollectiveWeightSynchronizer` gets attached. `nemo_rl/algorithms/ppo.py` builds its non-colocated NCCL group inline (`policy.init_collective(...)`, ppo.py:958-977) and calls `create_weight_synchronizer` only for the SGLang backend (ppo.py:948). So `refit_policy_generation` falls through to the legacy non-colocated path (grpo.py:2621), which never calls `offload_before_refit` — a PPO user who sets `policy.release_grads_before_refit: true` gets neither the memory release nor the `ValueError` from `validate_release_grads_before_refit`. The key is silently ignored.

Two things make this worth closing rather than deferring:

- PPO + Megatron + non-colocated vLLM is the supported topology and is nightly-tested — `tests/test_suites/llm/ppo-qwen2.5-1.5b-gsm8k-2n8g-megatron-valuetp2sp-dynbatch-noncolocated.sh` greps the run log for `"collective communication"`, i.e. asserts exactly that inline branch executes.
- Single-controller PPO *does* honor the flag, via `create_weight_synchronizer` at `nemo_rl/algorithms/single_controller_utils/setup.py:1502`, so the same algorithm behaves differently depending on the controller.

Relatedly, the new guide paragraph states "Unsupported combinations fail during synchronizer setup" (`docs/guides/refit.md:110`), which does not hold here because no synchronizer is built.

**Suggest fix:** either call `validate_release_grads_before_refit(...)` in `ppo.py::setup` so a misconfigured PPO run fails loudly, or add a PPO caveat next to the existing NIXL one at `docs/guides/refit.md:63-65`.

File: nemo_rl/weight_sync/factory.py [hunk 2]
```diff
@@ -84,6 +106,18 @@ def create_weight_synchronizer(
             f"Supported backends: {sorted(_SUPPORTED_BACKENDS)}"
         )
 
+    policy_cfg = getattr(policy, "cfg", {})
```
Comment (yuki-97) [inline nemo_rl/weight_sync/factory.py]: `getattr(policy, "cfg", {})` is defensiveness that no production path can reach. Every call site passes a `Policy` or `TQPolicy` (`grpo.py:1345/1774/1824/1842`, `ppo.py:948`, `distillation.py:600`, `single_controller_utils/setup.py:1336/1502`, `tools/refit_verifier.py:672`), and `Policy.__init__` sets `self.cfg` at `nemo_rl/models/policy/lm_policy.py:408`. This same function already dereferences `policy.cfg["megatron_cfg"]["enabled"]` unconditionally on the SGLang path below.

The consequence is silent: a future policy type that stores its config elsewhere would get `release_grads_before_refit=False` and skip `validate_release_grads_before_refit` entirely, rather than failing loudly.

The only thing exercising the fallback is `class LegacyPolicy: pass` in `test_missing_policy_config_defaults_refit_memory_release_to_disabled`, which models a type that does not exist. That test can go — the sibling `test_disabled_refit_memory_release_allows_missing_megatron_config` still covers the genuinely `NotRequired` `megatron_cfg` case, which the `(policy_cfg.get("megatron_cfg") or {})` guard on the next line handles.

```suggestion
    policy_cfg = policy.cfg
```

File: nemo_rl/models/policy/__init__.py [hunk 3]
```diff
@@ -600,6 +600,13 @@ class PolicyConfig(TypedDict):
...
+    # Move optimizer state to CPU when offload_before_refit runs. Defaults to True
+    # in the Megatron runtime for backward compatibility.
+    offload_optimizer_for_refit: NotRequired[bool]
```
Comment (yuki-97) [inline nemo_rl/models/policy/__init__.py]: This PR promotes `offload_optimizer_for_refit` into `PolicyConfig` and into the exemplar YAMLs with `true`, but its sole consumer at `nemo_rl/models/megatron/setup.py:427` still reads it as `bool(config.get("offload_optimizer_for_refit", True))`. Before this PR the default lived only in code; now it is stated in two places and can drift — flipping the exemplar to `false` would still leave `True` for any config that does not carry the key.

Per `.claude/skills/config-conventions/SKILL.md`, for a v1 `TypedDict` key the default lives only in the exemplar YAML and the call site reads it without a fallback; `.get(k, True)` is listed verbatim as a forbidden hidden default. The immediate neighbour is already in that shape — `offload_optimizer_for_logprob = config["offload_optimizer_for_logprob"]` at setup.py:426. The other key this PR adds follows the convention too (`policy_config.get("release_grads_before_refit") is True`).

If keeping the fallback is deliberate, so that pre-existing user YAMLs without the key still get `True`, that is reasonable — in that case a one-line comment at setup.py:427 saying the `True` is backcompat-only and the exemplar YAML is authoritative would save the next reader from reconciling the two.

File: nemo_rl/algorithms/distillation.py
Comment (yuki-97) [inline nemo_rl/algorithms/distillation.py]: On `nemo_rl/algorithms/distillation.py:900-905` — unchanged by this PR, hence a file-level comment: this comment becomes wrong once the new flag is on.

> The non-colocated refit path doesn't offload the student optimizer (offload_before_refit only runs in the colocated/Megatron path), so it's still on the train GPUs from the previous training step.

With `release_grads_before_refit: true` the refit routes through `CollectiveWeightSynchronizer.sync_weights`, which now calls `self._policy.offload_before_refit()` itself. So the non-colocated path *does* offload, and this call becomes the second `offload_before_refit()` on the student in the same step — nothing reloads between the refit at line 790 and here, since `prepare_for_training()` is at line 929.

The repeat itself is harmless (`move_optimizer("cpu")` skips non-CUDA tensors; `offload_to_cpu` is guarded on `storage().size() > 0`), so the only real issue is the stale parenthetical. Suggested wording:

```python
                        # The non-colocated refit path doesn't offload the student
                        # optimizer by default (only the colocated/Megatron path does,
                        # or non-colocated with release_grads_before_refit), so it may
                        # still be on the train GPUs from the previous training step.
                        # Offload it so the teacher fits for top-k inference;
                        # prepare_for_training() below reloads it.
```

File: nemo_rl/weight_sync/factory.py [hunk 4]
```diff
@@ -38,6 +38,28 @@
...
+    refit_transport: Optional[str],
+) -> None:
+    """Validate the topology supported by trainer memory release."""
+    if not enabled:
+        return
+    if not megatron_enabled:
+        raise ValueError(
+            "release_grads_before_refit requires the Megatron policy backend."
+        )
+    if colocated or generation_backend != VLLM_BACKEND or refit_transport is not None:
```
Comment (vasunvidia) [inline nemo_rl/weight_sync/factory.py]: Could release_grads_before_refit support be extended for nccl_reshard as well?

Comment (seonjinn) [reply nemo_rl/weight_sync/factory.py]: Sure, thank you for comment!

Comment (seonjinn) [reply nemo_rl/weight_sync/factory.py]: Implemented in b4e342c90. `release_grads_before_refit` now supports non-colocated vLLM `nccl_reshard`: the factory propagates the flag and the synchronizer runs `offload_before_refit()` immediately before each reshard. I also added validation, propagation, and ordering coverage, updated the refit guide, and merged the latest `main`.

**Comment by seonjinn** [comment]
/ok to test [071d219](https://github.com/NVIDIA-NeMo/RL/pull/3984/commits/071d2193f1393b0fb7beaa7cd1e25b1a36cea610)

**Comment by seonjinn** [comment]
/ok to test [5586e40](https://github.com/NVIDIA-NeMo/RL/pull/3984/commits/5586e408dd864c7e5ced954cb869da782103369e)

File: nemo_rl/weight_sync/factory.py [hunk 5]
```diff
@@ -38,6 +38,33 @@
...
+    refit_transport: Optional[str],
+) -> None:
+    """Validate the topology supported by trainer memory release."""
+    if not enabled:
+        return
+    if not megatron_enabled:
+        raise ValueError(
+            "release_grads_before_refit requires the Megatron policy backend."
+        )
+    if colocated or generation_backend != VLLM_BACKEND:
```
Comment (yuki-97) [inline nemo_rl/weight_sync/factory.py]: This adds a second switch for behaviour main already gates with `policy.generation.mcore_generation_config.offload_policy_before_refit` (#3300). Both call `policy.offload_before_refit()` before a non-colocated refit, so the key a user sets now depends on the generation backend, and line 56 rejects this one on the Megatron-generation path.

The trainer-side call is identical: [megatron_weight_synchronizer.py:167](https://github.com/NVIDIA-NeMo/RL/blob/5586e408dd864c7e5ced954cb869da782103369e/nemo_rl/weight_sync/megatron_weight_synchronizer.py#L167) vs [collective_weight_synchronizer.py:133](https://github.com/NVIDIA-NeMo/RL/blob/5586e408dd864c7e5ced954cb869da782103369e/nemo_rl/weight_sync/collective_weight_synchronizer.py#L133) and [nccl_reshard_weight_synchronizer.py:198](https://github.com/NVIDIA-NeMo/RL/blob/5586e408dd864c7e5ced954cb869da782103369e/nemo_rl/weight_sync/nccl_reshard_weight_synchronizer.py#L198). The name also undersells what runs: `offload_before_refit` also moves the optimizer to CPU and clears FP8 caches, which #3300's name describes.

**Suggest fix:**

1. Rename this key to `offload_policy_before_refit` so both backends share one spelling.
2. Optionally, have the Megatron-generation path read the same policy-level key and drop the `mcore_generation_config` copy, so one key covers all three non-colocated synchronizers.
